### ISSUE TITLE
Tried two levels of transclusion it did not work

### ISSUE THREAD

**Issue by michaelsjackson** [nobiot/org-transclusion#110]
Tried if I can transclude on two levels, but unfortunately it did not work. It seems currently only one level of transclusion is allowed. It does not have to be infinitely recursive but having at least a few levels of transclusion could be useful in practice.

For me a two level transclusion could be already a useful step. The two level transclusion should work, from any level you use transclusion.

Example 1:
A.org  < transcludes    B.org  < transcludes   C.org, D.org

C and D would be contained in B. B (together with C and D) would be contained in A.
Thus, triggering transclusion system from A.org would collect until C and D.

Example 2:
B.org  < transcludes    C.org   < transcludes   E.org, F.org
Same files as above, but now triggering from B, transclusion (two levels far) would contain this time also E and F inside C.

Not sure if easy to achieve, difficult or already available. Thanks for developing and sharing org-transclusion. This is very useful together with "header hiding technique" via :ignore: tags (another package).

**Comment by nobiot** [comment]
Thank you.

You could manually transclude C and D (in your Example 1) into B first, and then transclude B to A, then A should contain all of B, C and D. That's what  I have seen so far. 

```
A.org < transcludes B.org < transcludes C.org, D.org 
Depth 0 < Depth 1 < Depth 2
```

I must say, though, this outcome is not by design but by accident -- I could not foresee it. But it seems to work OK. As far as I could see, the problems are:

1. Live-sync edit does not work -- it does not break any of the involved buffers but the error message is not really instructive
2. Tranclusion of buffers depth 2 and onward is not done automatically from depth 0
3. Following from point 2, removing transclusions deeper than 1 from depth 0 does not work; probably, it should not

This is the current state.  I don't have good mental  capacity to achieve 2 and 3 properly at the moment. But I'm not against the idea.

What is your use case?

> This is very useful together with "header hiding technique" via :ignore: tags (another package).
Also what is the package?

**Comment by michaelsjackson** [comment]
Regarding the ignore headline technique, I am pasting a decription I found in stackoverflow, to answer your question in an understandable way:

   Add the following to your .emacs file:

    (require 'ox-extra)
    (ox-extras-activate '(ignore-headlines))

    Use the ignore tag on headlines you'd like to have ignored (while not ignoring their content).


Back to my writing. I like the above structuring technique while writing "more complex documents" so you do not lose the overview, you can restructure your document and ideas easiert, but the structure is invisible in the background (during export to latex-pdf). Probably great feature for anyone generating pdf documents from org-mode.

Now combine this "skelett-and-flesh" technique how I call it (the structure is the invisible skelett, and the content sentences is its visible flesh) with your transclusion and you have a quite powerful system of generating larger document from previous planning or written content. 

I was checking your transclusion, how far it could go, in the longer context and first obvious question coming to my mind was is it working at least 2 or 3 levels recursively. Why? Well, having no strong arguments now as use cases, but why not? Transclusions idea is bringing content from elsewhere to where I am now. Why should this only work over a "single bridge" and no multiple bridges?

How are you using your transclusion system? Only for one level depth?

**Comment by nobiot** [comment]
> How are you using your transclusion system? Only for one level depth?

Yes, only one level. I wrote the package to help me write a book on my professional domain (published in 2021). It's nothing related to programming or Emacs; I'm not a programmer by trade :). The analogy I had in mind was index cards. You write down a chunk of ideas or notes on a reference for each card; you arrange them to organise your thoughts. I think it's the physical version of what you call the "skelett-and-flesh" technique. That's what I wanted to have in Emacs to help me write the book.

> first obvious question coming to my mind was is it working at least 2 or 3 levels recursively. Why? Well, having no strong arguments now as use cases, but why not? Transclusions idea is bringing content from elsewhere to where I am now. Why should this only work over a "single bridge" and no multiple bridges?

For me, multiple depths of transclusion is difficult to understand, both practically for book-writing and technically for implementing in Elisp. 

This does not mean it should not be done. I may well be trapped in the analogy of index cards -- you probably wouldn't want to "transclude" content from another card, which would be just a "Refer to card No 221" rather than the actual content.

I think the real constraint is my technical ability. Once you go beyond depth 1, there is a case of A to B and B to A. In the case of Org-transclusion, A and B are not files; they are elements within a file. To me it's even harder (more tedious) to check these cyclic cases. And it's only one of the additional problems I have discovered (I have tried more than depth 1 in the course of development).

I'd be happy to get more technical advice or PR on this matter. Practically, I don't think I am capable of solving this problem in a practical way that can be implemented into this package.

**Comment by nobiot** [comment]
Also, thank you for answering the question. (ox-extras-activate '(ignore-headlines)) is indeed interesting. I didn't know about the package.

For what it's worth, Org-transclusion could do this already, if I didn't misunderstand what your requirement. 

> ‘:only-contents’
This property lets you exclude titles of headlines when you transclude a subtree (headline); you transclude only the contents. When the subtree contains sub-headlines, all the contents will be transcluded.

https://nobiot.github.io/org-transclusion/#Filter-Org-elements-per-transclusion

**Comment by lshir** [comment]
I know it has been a while, but I am interested in this use case so I will add some of my experience/thoughts.

I am using transclusion with `org-roam` for some writing. I write fragments in my dailies and aggregate/organize in the main org-roam knowledge base. (I would sort of organically stitch together bits of writing into sections/chapters by transclusion, and then aggregate further to see how it comes together.) 

I might have some fragments in sequence A, B, C as described by the original post above; initially I would start each note with a transclusion to the note before to get the whole text transitively.  It seemed to me that this was a misuse of the package and maybe not fully supported; I did run into some [infinite loop freezes](https://github.com/nobiot/org-transclusion/issues/110), so I stopped using this. Instead I created a main index/table of contents set of transclusions to avoid transitivity. I did find that overall transitivity worked better when I used `:only-contents` (in other cases, I would only get the header).

I am sorry this is a little unscientific but I bring this up because:
- I wonder what the current state/design of transitive inclusion today is (e.g. supported, sort of works but unsupported, won't be officially part of design, etc.)
  - and I'd be willing to put in a bit of programming if it's just something that you haven't gotten around to
- More broadly, as I was testing out the "right way" to stitch together different transclusions, I was curious if you have examples/best practices for organizing larger (writing) projects with this package. It would be extremely helpful to see what the intended larger-scale use case is like. 

Thanks so much!

**Comment by nobiot** [comment]
> I am sorry this is a little unscientific but I bring this up because:

Thank you for the comment. Please try not to worry about being "unscientific"; I find your writing clear. 

> * I wonder what the current state/design of transitive inclusion today is (e.g. supported, sort of works but unsupported, won't be officially part of design, etc.)

The design and implementation for recursive transclusion have not changed. Here is [my comment](https://github.com/nobiot/org-transclusion/issues/110#issuecomment-1002300638):

> I must say, though, this outcome is not by design but by accident -- I could not foresee it. But it seems to work OK. As far as I could see, the problems are:

The issue seems to me is `org-transclusion` directly works on the
source buffer. If the source buffer happens to have other transclusions,
they are treated as the buffer content. I am thinking that this was a
bad design. I probably should have created a temporary buffer and loaded
the file content (it can only contain the link to the transclusion
source, not its text). This way, this "accidental" recursion would not
occur.

It may sound like a reduced functionality. I think it may lead to finding a way to supporting some recursion. It would a bit tedious coding in my head:

1. Load the 1st transclusion content to a temp-buffer.
2. Search transclusion in 1.
3. If any found, repeat 1.
4. All the while, you would need to keep track of recursion, probably in an alist, so that you can see traverse the recursion tree backwards in a correct sequence.
5. If you find any transclusion link back to any of the file already in the recursion tree, do not transclude it. We need to avoid infinite loop -- the structure must be a tree. A circular network structure cannot be supported.
6. If this is done, live-edit must be disabled.

If you wish to start coding some bits of this, I'd love to the outcome :)


> * More broadly, as I was testing out the "right way" to stitch together different transclusions, I was curious if you have examples/best practices for organizing larger (writing) projects with this package. It would be extremely helpful to see what the intended larger-scale use case is like.

There is no "right way" in my view. Your way is the right way. If the tool does not let you, you will refine your toolset -- this is my way of doing things, and that's probably why I am using and coding Emacs.

As for me, I don't use `org-transclusion` much these days. When I was writing a book (https://github.com/nobiot/org-transclusion/issues/110#issuecomment-1004403657), I only had one-level flat transclusion and only a few of them. It was a good experiment for me. There seem to be many other people using `org-transclusion` in the way I didn't foresee in the beginning. So I maintain it. I would love to hear from others how they use it in their workflow.

**Comment by bradmont** [comment]
Hey, jumping in here since I ran into this today. I'm writing my PhD thesis using org-roam; I have piles of notes for sources, citations, book summaries, and so on. I also have notes for key ideas that I can quickly link to from sources, so when I want to write on an idea I open that org file, check the backlinks and write. Essentially I am cobbling together my thesis as a collection of notes, each with a "final version" header that I'll have included in the main document. But if in my reading I discover something that sheds new light on an idea I've already written on, it becomes quite easy to go and make the necessary changes to that section -- it's encapsulated in a relatively short org file. Complicated topics, like my research methodology, have a lot of connected ideas, so that file will be pulling several other notes, and then get pulled into the org file for the chapter, which will finally be pulled into the overall thesis org file before exporting with LaTeX.

Anyway, it's not super urgent, I can find an alternative workflow -- I think org's built-in #+includes can be nested, which is good enough for me. But this is how I would use nested or transitive transclusions.
